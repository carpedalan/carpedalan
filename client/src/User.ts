export enum User {
  read = 'read',
  write = 'write',
  none = 'none',
}
