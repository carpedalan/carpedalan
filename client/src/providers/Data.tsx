import * as React from 'react';

const { createContext, Children, useState } = React;
export interface Data {
  posts?: Components.Schemas.PostList;
}
const data: Data = {};
type SetData = (u: Data) => void;
export interface DataContext {
  data: Data;
  setData: SetData;
}
export const DataContext = createContext<DataContext>({
  data,
  setData: () => {},
});

export const DataConsumer = DataContext.Consumer;

export const DataProvider: React.FC<React.ReactNode> = ({ children }) => {
  const [dataState, setData] = useState<Data>(data);
  const setGlobalData = (arg: Data) => {
    setData({ ...dataState, ...arg });
  };
  const value: DataContext = {
    setData: setGlobalData,
    data: dataState,
  };

  return (
    <DataContext.Provider value={value}>
      {Children.only(children)}
    </DataContext.Provider>
  );
};
