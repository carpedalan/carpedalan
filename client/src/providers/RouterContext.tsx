import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { withRouter } from 'react-router-dom';

const { useContext } = React;

export const RouterContext = React.createContext({
  match: {
    isExact: false,
    params: {},
    path: '',
    url: '',
  },
  location: {
    pathname: '',
    search: {},
    state: {},
    hash: '',
  },
  history: {},
});

export const useRouterContext = () => useContext(RouterContext);

interface RouterContextProviderInterface extends RouteComponentProps {
  children: React.ReactChild;
}

export const RouterContextProvider = ({
  children,
  match,
  location,
  history,
}: RouterContextProviderInterface) => (
  <RouterContext.Provider value={{ match, location, history }}>
    {children}
  </RouterContext.Provider>
);

export default withRouter(RouterContextProvider);
