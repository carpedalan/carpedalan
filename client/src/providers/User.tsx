import * as React from 'react';

export type User = 'read' | 'write' | undefined;
const user: User = window.__SESSION__ ? window.__SESSION__.user : undefined;
const { createContext, Children, useState } = React;

type SetUser = (u: User) => void;
export interface IUserContext {
  user: User;
  setUser: SetUser;
}
export const UserContext = createContext<IUserContext>({
  user,
  setUser: () => {},
});

export const UserConsumer = UserContext.Consumer;

export const UserProvider: React.FC<React.ReactNode> = ({ children }) => {
  const [userState, setUser] = useState<User>(user);
  const value: IUserContext = {
    setUser,
    user: userState,
  };
  return (
    <UserContext.Provider value={value}>
      {Children.only(children)}
    </UserContext.Provider>
  );
};
