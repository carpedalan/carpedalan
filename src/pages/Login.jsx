import React, { useContext, useState } from 'react';
import { Redirect } from 'react-router-dom';
import styled from 'styled-components';
import request from 'superagent';

import InputField from '../fields/InputField';
import Field from '../form/Field';
import Form from '../form/Form';
import Submit from '../form/Submit';

import { User } from '..';

import ComingSoon from '../components/ComingSoon';
import Title from '../styles/Title';

const InputWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
`;
const InputForm = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  width: 25em;
  height: 25em;
`;

export default function Login() {
  const { setUser, user } = useContext(User);
  const [hasLoggedIn, setHasLoggedIn] = useState(false);
  const [showLogin, setShowLogin] = useState(false);
  const submitLogin = async ({ password }) => {
    try {
      const response = await request.post('/api/login', { password });
      setUser(response.body.user);
      setHasLoggedIn(true);
    } catch (e) {
      throw e;
    }
  };

  return (
    <>
      {showLogin || user ? (
        <InputWrapper>
          <InputForm>
            <Title center>Login</Title>
            <Form onSubmit={submitLogin}>
              <Field
                name="password"
                component={InputField}
                placeholder="Password, please"
                type="password"
              />
              <Submit />
            </Form>
          </InputForm>

          {hasLoggedIn ? <Redirect to="/" /> : null}
        </InputWrapper>
      ) : (
        <ComingSoon setShowLogin={setShowLogin} />
      )}
    </>
  );
}
