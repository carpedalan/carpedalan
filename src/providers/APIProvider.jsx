import React, { createContext, useState } from 'react';
import { node } from 'prop-types';
import styled from 'styled-components';
import request from 'superagent';

import { getThemeValue, DANGER_COLOR } from '../styles';

export const API = createContext({
  request: () => {},
  retry: () => {},
});

class APIError extends Error {}

const Toast = styled.div`
  position: fixed;
  bottom: 1em;
  width: 95vw;
  margin-left: 2.5vw;
  padding: 2em;
  background-color: ${getThemeValue(DANGER_COLOR)};
`;

export default function APIProvider({ children }) {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const patch = (...args) => {
    setLoading(true);
    const req = request.patch(...args);
    setLoading(false);
    return req;
  };
  const get = async (...args) => {
    let req;
    try {
      setLoading(true);
      req = await request.get(...args);
      if (!req.body) throw new APIError('shit!');
    } catch (e) {
      setError('Get failed');
    } finally {
      setLoading(false);
    }
    return req;
  };

  const post = (...args) => {
    setLoading(true);
    const req = request.post(...args);
    setLoading(false);
    return req;
  };
  const del = (...args) => {
    setLoading(true);
    const req = request.del(...args);
    setLoading(false);
    return req;
  };

  return (
    <API.Provider value={{ patch, get, post, del, loading }}>
      {children}
      {error ? <Toast>{error}</Toast> : null}
    </API.Provider>
  );
}

APIProvider.propTypes = {
  children: node.isRequired,
};
