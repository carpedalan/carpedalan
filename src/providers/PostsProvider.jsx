import React, { createContext, useState } from 'react';
import { node } from 'prop-types';
import { stringify } from 'qs';
import { CellMeasurerCache } from 'react-virtualized';
import request from 'superagent';

import { API_PATH } from '../../shared/constants';
import log from '../utils/log';
import { performance } from '../utils/globals';

import addPlaceholderColor from './postUtils';

export const Posts = createContext({
  posts: [],
  getPosts: () => {},
  meta: {},
});

const PostProvider = ({ children }) => {
  const [posts, setPosts] = useState([]);
  const [meta, setMeta] = useState({ count: 0 });
  const [currentPage, setCurrentPage] = useState(1);
  const [progressMap, setProgressMap] = useState({});

  const cache = new CellMeasurerCache({
    fixedWidth: true,
    defaultHeight: 500,
  });

  const patchPost = id => async values => {
    try {
      const { body } = await request.patch(`${API_PATH}/posts/${id}`, values);
      setPosts(posts.map(post => (post.id === body.id ? body : post)));
    } catch (e) {
      log.error(e);
    }
  };

  const delPost = id => async () => {
    await request.delete(`${API_PATH}/posts/${id}`);
    setPosts(posts.filter(post => post.id !== id));
    cache.clearAll();
  };

  const getPosts = async (page = currentPage) => {
    try {
      const pageQuery = {
        page,
      };

      // Some super basic caching - don't refetch if we have already.
      if (posts[page * 100]) return null;
      const apiCall = request.get(`${API_PATH}/posts?${stringify(pageQuery)}`);
      const response = await apiCall;
      setPosts([...posts, ...response.body.data].map(addPlaceholderColor));
      setMeta(response.body.meta);
      setCurrentPage(currentPage + 1);
    } catch (e) {
      log.error('loading failed');
    }
    return null;
  };

  const invalidateAll = () => {
    cache.clearAll();
    setPosts([]);
    setMeta({ count: 0 });
  };

  const createPost = async (formData, index = 0) => {
    try {
      const timeBefore = performance.now();
      await new Promise(resolve => setTimeout(resolve, 2000));
      const response = await request
        .post(`${API_PATH}/posts`)
        .send(formData)
        .on('progress', e => {
          if (e.percent) {
            setProgressMap({ ...progressMap, [index]: e.percent });
            console.error('progress', e.percent); // eslint-disable-line
          }
        });
      const timeAfter = performance.now();
      invalidateAll();

      return { response: response.body, timeElapsed: timeAfter - timeBefore };
    } catch (e) {
      log.error(e);
      return Promise.reject(e);
    }
  };

  return (
    <Posts.Provider
      value={{
        cache,
        getPosts,
        posts,
        meta,
        patchPost,
        delPost,
        invalidateAll,
        createPost,
        progressMap,
      }}
    >
      {children}
    </Posts.Provider>
  );
};

PostProvider.propTypes = {
  children: node.isRequired,
};

export default PostProvider;
