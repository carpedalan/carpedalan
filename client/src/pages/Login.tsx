import useApi, { HttpMethods } from 'hooks/useApi';
import useForm from 'hooks/useForm';
import useUser from 'hooks/useUser';
import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { default as styled } from 'styled-components';
import Button from 'styles/Button';
import Input from 'styles/Input';
import { User } from 'User';
import { InputForm, InputWrapper, StyledButton, StyledTitle } from './styles';
/**
 * Login route component. Includes form and api effects
 */
export const Login: React.FC<RouteComponentProps> = ({ history }) => {
  const { setUser, user } = useUser();
  const loginPost = useApi('/login', HttpMethods.post);
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    return e.target.value;
  };
  const { useField, setValue, form } = useForm('myForm');
  const input = useField({ handleChange, field: 'password ' });
  const handleSubmit = async (
    e: React.FormEvent<HTMLFormElement>,
  ): Promise<void> => {
    e.preventDefault();
    try {
      console.log('form', form);
      const response = await loginPost.request({ password: form.password });
    } catch (e) {
      console.error('e');
    }
  };

  return (
    <InputWrapper>
      <InputForm onSubmit={handleSubmit}>
        <StyledTitle center={true}>Login</StyledTitle>
        {user}
        <Input>
          <input type="password" {...input} />
        </Input>
        <Button type="submit">Login</Button>
      </InputForm>
      {loginPost.response ? loginPost.response.data : null}
      {history.location.pathname}
    </InputWrapper>
  );
};

export default Login;
