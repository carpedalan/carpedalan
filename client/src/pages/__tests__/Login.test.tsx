import axios, { AxiosResponse, AxiosStatic } from 'axios';
import { shallow } from 'enzyme';
import { UserContext, UserProvider } from 'providers/User';
import * as React from 'react';
import { MemoryRouter, Redirect, Route } from 'react-router';
import { BrowserRouter, Link } from 'react-router-dom';
import {
  act,
  cleanup,
  fireEvent,
  render,
  waitForElement,
} from 'react-testing-library';

import Login from '../Login';

const mockedAxios = axios as jest.Mocked<typeof axios>;

jest.mock('axios');

describe('<Login />', () => {
  beforeEach(jest.resetAllMocks);
  it('should match snapshot', () => {
    const app = shallow(<Login />);
    expect(app).toMatchSnapshot();
  });

  it('should fire an api call on submit', () => {
    const app = shallow(<Login />);
    const val = 'val';
    app.find('[data-test="submit"]').simulate('submit', new Event('submit'));
    expect(mockedAxios.post).toHaveBeenCalledWith('/v1/login', {
      password: '',
    });
  });

  it('should have a working input', () => {
    const app = shallow(<Login />);
    const val = 'val';
    app
      .find('[data-test="password"]')
      .simulate('change', { target: { value: val } });
    expect(app.find('[data-test="password"]').props().value).toBe(val);
  });

  it('should show an error if there is one', async () => {
    mockedAxios.post.mockImplementation(() =>
      Promise.reject({ response: { data: { message: 'error' } }, status: 420 }),
    );
    const app = shallow(<Login />);
    app.find('[data-test="submit"]').simulate('submit', new Event('submit'));

    await new Promise(resolve => setTimeout(resolve, 0));

    expect(app.find('[data-test="error"]').text()).toBe('error');
  });

  it('should not show the error after typing after an error', async () => {
    mockedAxios.post.mockImplementation(() =>
      Promise.reject({ response: { data: { message: 'error' } }, status: 420 }),
    );
    const app = shallow(<Login />);
    app.find('[data-test="submit"]').simulate('submit', new Event('submit'));

    await new Promise(resolve => setTimeout(resolve, 0));
    expect(app.find('[data-test="error"]').text()).toBe('error');
    app
      .find('[data-test="password"]')
      .simulate('change', { target: { value: 'something' } });
    expect(app.find('[data-test="error"]').exists()).toBe(false);
  });

  it('should show an error', async () => {
    mockedAxios.post.mockImplementation(() =>
      Promise.reject({ response: { data: { message: 'error' } }, status: 420 }),
    );
    const { getByTestId } = render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
    );
    fireEvent.click(getByTestId('submit-button'));
    const errorText = await waitForElement(() => getByTestId('error'));

    expect(errorText).toHaveTextContent('error');
  });

  it('should redirect on success', async () => {
    const RedirectComponent = () => <div data-testid="redirected">Hi</div>;
    mockedAxios.post.mockImplementation(() =>
      Promise.resolve({
        data: { user: 'read' },
        status: 200,
      } as AxiosResponse),
    );
    const { getByTestId } = render(
      <UserProvider>
        <MemoryRouter>
          <Link data-testid="login-link" to="login">
            Login
          </Link>
          <Route path="/login" component={Login} />
          <UserContext.Consumer>
            {context => (context.user ? <div data-testid="home" /> : null)}
          </UserContext.Consumer>
        </MemoryRouter>
      </UserProvider>,
    );
    fireEvent.click(getByTestId('login-link'));
    const errorText = await waitForElement(() => getByTestId('submit'));

    fireEvent.click(getByTestId('submit-button'));
    const redirected = await waitForElement(() => getByTestId('home'));
  });
});
