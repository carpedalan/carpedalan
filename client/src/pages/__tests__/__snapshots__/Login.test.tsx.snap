// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should match snapshot 1`] = `
<styled.div>
  <styled.form
    data-test="submit"
    data-testid="submit"
    onSubmit={[Function]}
  >
    <Styled(styled.h1)
      center={true}
    >
      Logizn
    </Styled(styled.h1)>
    <styled.div>
      <input
        data-test="password"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </styled.div>
    <styled.button
      data-test="submit-button"
      data-testid="submit-button"
      type="submit"
    >
      Login
    </styled.button>
    <Link
      to="/request"
    >
      Request access
    </Link>
  </styled.form>
</styled.div>
`;
